rosbag2_storage/include/rosbag2_storage/serialized_bag_message.hpp

```diff
@@ -0,0 +1,32 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_STORAGE__SERIALIZED_BAG_MESSAGE_HPP_
+#define ROSBAG2_STORAGE__SERIALIZED_BAG_MESSAGE_HPP_
+
+#include "rcutils/types/char_array.h"
+#include "rcutils/time.h"
+
+namespace rosbag2_storage
+{
+
+struct SerializedBagMessage
+{
+  rcutils_char_array_t serialized_data;
```

Comment:
In principle, this is what I would've expected - except perhaps a `shared_ptr<rcutils_char_array_t>` because that seems to be the type we get in a subscription callback.

However, I have two questions:
a) The original design idea was to have librosbag2_storage independent of ros2. This part here would imply that we need at least a dependency on rcutils and/or rmw. Obviously, those are just type definitions, so it might be ok.
b) We tried to have a closer look and see how to write this to sqlite. It seems `rcutils_char_array_t` consists of four parts, the name-sake `char` array with the real content, two size bytes and another struct of allocator functions. I don't exactly understand the last part, the `rcutils_allocator_t`. We can't write it to a database (at least I wouldn't know how) and I don't understand how we can easily generate this part - can we always take the default from the rmw? How does this work in relation to part a) of my trouble?

Reply:
> a) The original design idea was to have librosbag2_storage independent of ros2. This part here would imply that we need at least a dependency on rcutils and/or rmw. Obviously, those are just type definitions, so it might be ok.

As much as possible it should be disconnected from ROS 2, to make it as portable as possible. If it's just `rcutils` then it shouldn't be too hard to do that, as `rcutils` doesn't really have any dependencies (no run dependencies, ament and python-empy as build depends). A dependency like `rmw` or `rosidl`, on the other hand, would be a red flag for me.

> We tried to have a closer look and see how to write this to sqlite. It seems `rcutils_char_array_t` consists of four parts, the name-sake char array with the real content, two size bytes and another struct of allocator functions. I don't exactly understand the last part, the `rcutils_allocator_t`. We can't write it to a database (at least I wouldn't know how) and I don't understand how we can easily generate this part - can we always take the default from the `rmw`? How does this work in relation to part a) of my trouble?

You don't need to store the allocator structure in the database. That's just used to resize the serialized message and later to "delete it". Since the function that writes it to the database could essentially take a `const rcutils_char_array_t *`, the allocator would not be used nor is it needed when bringing it out of the database later.

On the other side (reading a bag), some allocator will be used to create the space in the outgoing `rcutils_char_array_t` instance so that the data from the database can be copied there.

There's no need for `rmw` in any of these steps. The allocator type and all related functions are also part of `rcutils`.

----

The other alternative would be to provide duplicate data types for the character array and allocator in this library itself. Converting between the version here and `rcutils` should be straightforward, but it would be nice to reuse what's in `rctuils`.